### PR TITLE
Use floor() instead of round() in rest_is_integer()

### PR DESCRIPTION
**Pull request by siliconforks**
This fixes a problem with large integers causing `rest_is_integer()` to return `false`.

It also adds some tests for large integers and a few other things that did not have any existing tests (negative floats, scientific notation).

To run the test suite:

```
vendor/bin/phpunit --filter Tests_REST_API
```

Trac ticket: https://core.trac.wordpress.org/ticket/65271

## Use of AI Tools

AI assistance: Yes
Tool(s): Claude Opus 4.7
Used for: @westonruter said that Claude Opus 4.7 made suggestions to add separate branches for `int` values and `string` integers; these looked fine to me, so I added them.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/rest-api.php [hunk 1]
```diff
@@ -2349,6 +2357,47 @@ public function data_rest_is_integer() {
...
+				'15e+0',
+			),
+			array(
+				true,
+				'15e-0',
+			),
+			array(
+				false,
+				'15e-1',
+			)
```
Comment (westonruter) [inline tests/phpunit/tests/rest-api.php]:
```suggestion
			),
```

File: tests/phpunit/tests/rest-api.php [hunk 2]
```diff
@@ -2349,6 +2357,47 @@ public function data_rest_is_integer() {
...
+		 * These values failed with PHP 8.4 when rest_is_integer() used round() in its implementation.
+		 */
+		return array(
+			array(
+				2 ** 52,
+			),
+			array(
+				2 ** 52 + 2,
+			),
 		);
```
Comment (westonruter) [inline tests/phpunit/tests/rest-api.php]: Why not add this to the existing `data_rest_is_integer` data provider instead?

File: tests/phpunit/tests/rest-api.php (hunk 1 above)
Comment (westonruter) [inline tests/phpunit/tests/rest-api.php]: Let's also add some big int strings, like `'4503599627370498'` (2 ** 52 + 2) and `'4503599627370496'` (2 ** 52).

File: src/wp-includes/rest-api.php [hunk 3]
```diff
@@ -1570,7 +1570,17 @@ function rest_is_boolean( $maybe_bool ) {
...
+		return true;
+	}
+
+	// A canonical integer string of any magnitude — verified without float conversion.
+	if ( is_string( $maybe_integer ) && preg_match( '/^\s*[+-]?[0-9]+\s*$/', $maybe_integer ) ) {
+		return true;
+	}
+
+	// Decimal and scientific-notation strings (and floats) keep their historical behavior.
+	return is_numeric( $maybe_integer ) && floor( (float) $maybe_integer ) === (float) $maybe_integer;
```
Comment (westonruter) [inline src/wp-includes/rest-api.php]: This seems marginally better:

```suggestion
	if ( ! is_numeric( $maybe_integer ) ) {
		return false;
	}
	$float_value = (float) $maybe_integer;
	return floor( $float_value ) === $float_value;
```

File: tests/phpunit/tests/rest-api.php
Comment (westonruter) [inline tests/phpunit/tests/rest-api.php]:
```suggestion
	/**
	 * Data provider for {@see self::test_rest_is_integer()}.
	 * 
	 * @return list<array{ bool, mixed }>
	 */
	public function data_rest_is_integer(): array {
```

File: tests/phpunit/tests/rest-api.php [hunk 4]
```diff
@@ -2349,6 +2358,42 @@ public function data_rest_is_integer() {
...
+				'15e-0',
+			),
+			array(
+				false,
+				'15e-1',
+			),
+
+			/*
+			 * These values failed with PHP 8.4 when rest_is_integer() used round() in its implementation.
+			 */
```
Comment (westonruter) [inline tests/phpunit/tests/rest-api.php:2431]:
```suggestion
			// The following values failed with PHP 8.4 when rest_is_integer() used round() in its implementation.
```

File: src/wp-includes/rest-api.php [hunk 5]
```diff
@@ -1570,7 +1570,21 @@ function rest_is_boolean( $maybe_bool ) {
  * @return bool True if an integer, otherwise false.
  */
 function rest_is_integer( $maybe_integer ) {
```
Comment (westonruter) [inline src/wp-includes/rest-api.php]: Since we're here!

```suggestion
function rest_is_integer( $maybe_integer ): bool {
```

**Review by westonruter — APPROVED** [review]

File: tests/phpunit/tests/rest-api.php [hunk 6]
```diff
@@ -2349,6 +2396,120 @@ public function data_rest_is_integer() {
...
+			array(
+				false,
+				false,
+			),
+			array(
+				false,
+				null,
+			),
+
+			// The following values failed with PHP 8.4 when rest_is_integer() used round() in its implementation.
```
Comment (siliconforks) [inline tests/phpunit/tests/rest-api.php]: It's a bit confusing now what values this comment actually applies to.  Originally there were 4 values following the comment, and the comment was intended to refer to all of them.  You (and Claude) added a 5th and a 6th value in 7a07f03d395047a0ec5a49ea45bb7cac044a9844.  The comment still makes sense for the 5th value (`'-4503599627370498'`).  However, it doesn't make sense for the 6th value (`'4611686018427387904'`) - that value would be correctly identified as an integer even with the old implementation that used `round()`.

You could just delete the comment; it's not critical information, although I think it is useful to point out the kinds of values that weren't working with the old implementation.

**Comment by coderGtm** [comment]
@westonruter I beleive this PR logically solves the problem in hand. Although my personal preference would still be a more simpler and straightforward regex approach like in #11883. That is predictable and dependency free (core php changes don't affect the behaviour of the function). It's also more maintainable and easy to extend/change the ruleset (if ever needed).

I'd also propose to not let values pass which should not be passed like unrealistic scientific notations, since I think it was an unintended side effect of the original implementation.

That being said, I'm open to any approach the community decides upon. I'm just sharing my opinion.

File: tests/phpunit/tests/rest-api.php (hunk 6 above)
Comment (westonruter) [reply tests/phpunit/tests/rest-api.php]: How about something like:


```suggestion
			// The following values test very large integers.
```

**Comment by westonruter** [comment]
@coderGtm values may be unexpected, but at the scale of WordPress, they'll surely happen! Backwards-compatibility is key.

File: tests/phpunit/tests/rest-api.php (hunk 6 above)
Comment (siliconforks) [reply tests/phpunit/tests/rest-api.php]: That seems fine.
